x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_exploration/components/total_feature_importance_summary/feature_importance_summary.tsx

```diff
@@ -0,0 +1,257 @@
+        {
+          defaultMessage: 'magnitude',
+        }
+      ) as string,
+    };
+    let classificationType:
+      | 'binary_classification'
+      | 'multiclass_classification'
+      | 'regression'
+      | '' = '';
+    if (totalFeatureImportance.length < 1) {
+      return [sortedData, _barSeriesSpec];
+    }
+
+    if (isClassificationTotalFeatureImportance(totalFeatureImportance[0])) {
+      // if binary classification
+      if (totalFeatureImportance[0].classes.length === 2) {
+        classificationType = 'binary_classification';
+        sortedData = (totalFeatureImportance as ClassificationTotalFeatureImportance[])
+          .map((d) => {
+            return {
+              featureName: d.feature_name,
+              // in case of binary classification, both classes will have the same mean importance
+              meanImportance: d.classes[0].importance.mean_magnitude,
+            };
+          })
+          .sort((a, b) => b.meanImportance - a.meanImportance);
+      }
+
+      // if multiclass classification
+      // stack them in order of increasing importance
+      // so for each feature, biggest importance on the left to smallest importance on the right
+      if (totalFeatureImportance[0].classes.length > 2) {
+        classificationType = 'multiclass_classification';
+
+        (totalFeatureImportance as ClassificationTotalFeatureImportance[])
+          .sort(
+            (prevFeature, currentFeature) =>
+              calculateTotalMeanImportance(currentFeature) -
+              calculateTotalMeanImportance(prevFeature)
+          )
+          .forEach((feature) => {
+            const sortedFeatureClass = feature.classes.sort(
+              (a, b) => b.importance.mean_magnitude - a.importance.mean_magnitude
+            );
+            sortedData.push(
+              ...sortedFeatureClass.map((featureClass) => ({
+                featureName: feature.feature_name,
+                meanImportance: featureClass.importance.mean_magnitude,
+                className: featureClass.class_name,
+              }))
+            );
+          });
+
+        _barSeriesSpec = {
+          xAccessor: 'featureName',
+          yAccessors: ['meanImportance'],
+          splitSeriesAccessors: ['className'],
+          stackAccessors: ['featureName'],
+        };
+      }
+    }
+    // if regression
+    if (isRegressionTotalFeatureImportance(totalFeatureImportance[0])) {
+      classificationType = 'regression';
+
+      sortedData = (totalFeatureImportance as RegressionTotalFeatureImportance[])
+        .map((d: RegressionTotalFeatureImportance) => ({
+          featureName: d.feature_name,
+          meanImportance: d.importance.mean_magnitude,
+        }))
+        .sort((a, b) => b.meanImportance - a.meanImportance);
+    }
+
+    // only show legend if it's a multiclass
+    const _showLegend = classificationType === 'multiclass_classification';
```

Comment:
For classification jobs where the numbers of classes is higher than the number of features, I am seeing some truncation of the legend labels. Can we fix this with styling on our side, or does it need a fix on the chart?

![image](https://user-images.githubusercontent.com/7405507/94674328-e9647600-030f-11eb-9ac1-db246956a189.png)

I also see this with a job on the mushroom data set with this config:

```
{
  "id": "mushroom_cap_color_classification2",
  "description": "10 top feature importance values",
  "source": {
    "index": [
      "mushroom"
    ],
    "query": {
      "match_all": {}
    }
  },
  "dest": {
    "index": "mushroom_cap_color_classification2",
    "results_field": "ml"
  },
  "analysis": {
    "classification": {
      "dependent_variable": "cap-color",
      "num_top_feature_importance_values": 10,
      "class_assignment_objective": "maximize_minimum_recall",
      "num_top_classes": 2,
      "prediction_field_name": "cap-color_prediction",
      "training_percent": 8,
      "randomize_seed": 2195820625972847400
    }
  },
  "analyzed_fields": {
    "includes": [
      "bruises",
      "cap-color",
      "cap-shape",
      "cap-surface",
      "edibility",
      "gill-attachment",
      "gill-color",
      "gill-size",
      "gill-spacing",
      "habitat",
      "odor",
      "population",
      "ring-number",
      "ring-type",
      "spore-print-color",
      "stalk-color-above-ring",
      "stalk-color-below-ring",
      "stalk-root",
      "stalk-shape",
      "stalk-surface-above-ring",
      "stalk-surface-below-ring",
      "veil-color",
      "veil-type"
    ],
    "excludes": []
  },
  "model_memory_limit": "41mb",
  "create_time": 1601285932889,
  "version": "8.0.0",
  "allow_lazy_start": false,
  "max_num_threads": 1
}
```

Reply:
Updated it to give some buffer spacing here https://github.com/elastic/kibana/pull/78238/commits/17cfd1d0bb70eac6857c3bc8e6895f702c77c791

Unfortunately it's not dynamic so if the label texts are short, there will be some space on the right and vice versa it will still be truncated if the texts are long. Let me know if you're good with the change! Or if there's anything we can improve as far the legend goes, we can do it in a follow up pR.

Reply:
I think this is a good solution. Shame it isn't dynamic, but it fixes the issue I was seeing where it was truncating the labels after the first character. For long labels the majority of the text should be visible so the user can easily identify which color goes with each class.